### PR TITLE
V2 annotator comments

### LINKED ISSUE #994: Highlight text
I used wallabag, in a specific manner.
I read a lot of articles. _A LOT_.
I would like to keep track of them and the best would be if **I could highlight important phrases**.
For example, instapaper offers a highlighting solution. 

Is it possible to do the same in wallabag?

_+ We could do better by blurring the text wich is not highlight, if necessary_

### PR DESCRIPTION
**Pull request by tcitworld**
- [x] post comment
- [x] read comments for an entry
- [x] edit comment
- [x] delete comment
- [x] tests
- [x] Bring annotator to baggy
- [x] little documentation

For 2.0 stable :
- API stuff

For 2.1 :
-  Dealing with multiple users and all this stuff

Fix #994

### CODE REVIEW

**Comment by tcitworld** [comment]
Okay, the edit and delete functions are very discreet.

File: app/config/routing_rest.yml [hunk 1]
```diff
@@ -1,3 +1,3 @@
 Rest_Wallabag:
   type : rest
-  resource: "@WallabagApiBundle/Resources/config/routing_rest.yml"
+  resource: "@WallabagApiBundle/Resources/config/routing_rest.yml"
```
Comment (nicosomb) [inline app/config/routing_rest.yml]: Can you install editorconfig plugin on your favorite IDE please? ;-)

File: src/Wallabag/CommentBundle/Controller/WallabagCommentController.php [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+     *      requirements={
+     *          {"name"="entry", "dataType"="integer", "requirement"="\w+", "description"="The entry ID"}
+     *      }
+     * )
+     *
+     * @return Response
+     */
+    public function getAnnotationsAction(Entry $entry)
+    {
+        $commentrows = $this
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/Controller/WallabagCommentController.php]: `$commentRows`

File: src/Wallabag/CommentBundle/Controller/WallabagCommentController.php [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+     *
+     * @ApiDoc(
+     *      requirements={
+     *          {"name"="comment", "dataType"="string", "requirement"="\w+", "description"="The comment ID"}
+     *      }
+     * )
+     *
+     * @return Response
+     */
+    public function putAnnotationAction($idcomment, Request $request)
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/Controller/WallabagCommentController.php]: `$idComment`

File: src/Wallabag/CommentBundle/Controller/WallabagCommentController.php [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        $this->validateAuthentication();
+
+        $data = json_decode($request->getContent(), true);
+
+        $comment = $this
+                ->getDoctrine()
+                ->getRepository('WallabagCommentBundle:Comment')
+                ->findCommentById($idcomment);
+
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/Controller/WallabagCommentController.php]: And if you comment doesn't exist in database?

File: src/Wallabag/CommentBundle/Controller/WallabagCommentController.php [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+     * @return Response
+     */
+    public function deleteAnnotationAction($idcomment)
+    {
+        $this->validateAuthentication();
+
+        $comment = $this
+                ->getDoctrine()
+                ->getRepository('WallabagCommentBundle:Comment')
+                ->findCommentById($idcomment);
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/Controller/WallabagCommentController.php]: Same as previous: if you comment doesn't exist in database?

File: src/Wallabag/CommentBundle/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Wallabag\CommentBundle\DependencyInjection;
+
+use Symfony\Component\Config\Definition\Builder\TreeBuilder;
+use Symfony\Component\Config\Definition\ConfigurationInterface;
+
+/**
+ * This is the class that validates and merges configuration from your app/config files.
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/DependencyInjection/Configuration.php]: you can remove all the comments in this file.

File: src/Wallabag/CommentBundle/DependencyInjection/WallabagCommentExtension.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Wallabag\CommentBundle\DependencyInjection;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\HttpKernel\DependencyInjection\Extension;
+use Symfony\Component\DependencyInjection\Loader;
+
+/**
+ * This is the class that loads and manages your bundle configuration.
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/DependencyInjection/WallabagCommentExtension.php]: you can remove all the comments in this file.

File: src/Wallabag/CommentBundle/DataFixtures/ORM/LoadCommentData.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+     * {@inheritdoc}
+     */
+    public function load(ObjectManager $manager)
+    {
+        $comment1 = new Comment($this->getReference('admin-user'));
+        $comment1->setEntry($this->getReference('entry1'));
+        $comment1->setText('This is my comment /o/');
+        $comment1->setUpdated(new \DateTime());
+        $comment1->setQuote('content');
+        //$comment1->setRanges(json_encode([{"start":"","startOffset":24,"end":"","endOffset":31}]));
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/DataFixtures/ORM/LoadCommentData.php]: Why is this field commented?

File: src/Wallabag/CommentBundle/Entity/Comment.php [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        return $this->entry;
+    }
+
+    /**
+     * @VirtualProperty
+     * @SerializedName("annotator_schema_version")
+     */
+    public function getVersion()
+    {
+        return 'v1.0';
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/Entity/Comment.php]: :+1:

File: src/Wallabag/CommentBundle/Entity/Comment.php [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+     * @var \DateTime
+     *
+     * @ORM\Column(name="updated", type="datetime", nullable=true)
+     */
+    private $updated;
+
+    /**
+     * @var string
+     *
+     * @ORM\Column(name="quote", type="string", length=255, nullable=true)
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/Entity/Comment.php]: 255 is enough?

File: src/Wallabag/CommentBundle/Entity/Comment.php [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+     */
+    private $quote;
+
+    /**
+     * @var array
+     *
+     * @ORM\Column(name="ranges", type="array", length=255, nullable=true)
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/Entity/Comment.php]: 255 is enough?

File: app/config/routing_rest.yml (hunk 1 above)
Comment (tcitworld) [inline app/config/routing_rest.yml]: Well, it is. Strange.

File: src/Wallabag/CommentBundle/Controller/WallabagCommentController.php [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+     * @param string $json
+     *
+     * @return Response
+     */
+    private function renderJsonResponse($json, $code = 200)
+    {
+        return new Response($json, $code, array('application/json'));
+    }
+
+    private function validateAuthentication()
```
Comment (j0k3r) [inline src/Wallabag/CommentBundle/Controller/WallabagCommentController.php]: You should better add a new rule in the `security.yml` file related to your url instead.
https://github.com/wallabag/wallabag/blob/v2/app/config/security.yml#L54

File: src/Wallabag/CommentBundle/Entity/Comment.php (hunk 10 above)
Comment (j0k3r) [inline src/Wallabag/CommentBundle/Entity/Comment.php]: Also, I don't think it should be `nullable`

File: src/Wallabag/CommentBundle/Entity/Comment.php (hunk 11 above)
Comment (j0k3r) [inline src/Wallabag/CommentBundle/Entity/Comment.php]: Same here for `nullable`

File: src/Wallabag/CommentBundle/Entity/Comment.php [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+    }
+
+    /**
+     * Set created.
+     *
+     * @param \DateTime $created
+     *
+     * @return Comment
+     */
+    public function setCreated($created)
```
Comment (j0k3r) [inline src/Wallabag/CommentBundle/Entity/Comment.php]: Did you plan to use these methods?
I think you should better handle created / update the way we did in [Entry](https://github.com/wallabag/wallabag/blob/v2/src/Wallabag/CoreBundle/Entity/Entry.php#L355-L366).

File: src/Wallabag/CoreBundle/Resources/views/themes/material/Entry/entry.html.twig [hunk 14]
```diff
@@ -147,6 +147,41 @@ main {
...
+
+
+var app = new annotator.App();
+app.include(annotator.ui.main, {
+    element: document.querySelector('article')
+});
+app.include(annotator.storage.http, {
+    prefix: '/annotations',
+    urls: {
+        create: '/{{ entry.id }}',
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Resources/views/themes/material/Entry/entry.html.twig]: You should better generate each url using twig instead of _by hand_. It avoids to use a prod url in a dev environment (since your are calling `/annotations`)

**Comment by j0k3r** [comment]
Do you have an idea on why the default called url when opening an article contains a parameter with entry id: `/annotations/7?entry=7` ?

**Comment by tcitworld** [comment]
I added https://github.com/wallabag/wallabag/pull/1653/files#diff-2c1424825ef3119eace5cdec44e06b10R153 but it may not be needed and therefore causing this.

**Comment by j0k3r** [comment]
Yep we don't need that

**Comment by j0k3r** [comment]
BTW, if we didn't merge that for the first beta we'll need to generate a Doctrine migration class

**Comment by j0k3r** [comment]
What about returning comments in the entry api ?

**Comment by tcitworld** [comment]
Will do.

**Comment by tcitworld** [comment]
Comments are indeed returned for an entry :
```
"comments": [
                    {
                        "annotator_schema_version": "v1.0",
                        "created_at": "2016-02-20T11:16:12+0000",
                        "id": 1,
                        "quote": "Samsung",
                        "ranges": [
                            {
                                "end": "/div[1]/main[1]/div[1]/div[1]/article[1]/p[5]",
                                "endOffset": 77,
                                "start": "/div[1]/main[1]/div[1]/div[1]/article[1]/p[5]",
                                "startOffset": 70
                            }
                        ],
                        "text": "Coréen",
                        "updated_at": "2016-02-20T11:16:12+0000"
                    }
                ],
```

**Comment by jlnostr** [comment]
Nice, I might use this feature a lot :+1:

**Comment by tcitworld** [comment]
> Nice, I might use this feature a lot :+1:

Posting and editing comments API's auth isn't ready, though.

**Comment by jlnostr** [comment]
> Posting and editing comments API's auth isn't ready, though.

I can wait :)

**Comment by nicosomb** [comment]
Do we need a parameter to enable/disable this feature for each user? 

As a user, maybe I just want to copy/paste text in an article and I don't want to see the annotator icon.

**Comment by tcitworld** [comment]
> As a user, maybe I just want to copy/paste text in an article and I don't want to see the annotator icon.

It doesn't shock me.

**Comment by j0k3r** [comment]
We can still provide an option (enable by default) in the future to disable that. We'll if people complains.

**Comment by nicosomb** [comment]
OK. So we can rebase and merge.

**Comment by j0k3r** [comment]
> What about returning comments in the entry api ?

So this is planned for later?

**Comment by tcitworld** [comment]
> So this is planned for later?

It's already there as I've shown above.

**Comment by j0k3r** [comment]
:grin: :+1: _sorry_

**Comment by tcitworld** [comment]
Before we merge that, there's two more things to do (see first post).

**Comment by nicosomb** [comment]
> Displaying how much comments there are for an entry on top of it.

Can be done after this PR.

**Comment by nicosomb** [comment]
I added a new todo: documentation.

**Comment by j0k3r** [comment]
Rebased :white_check_mark:

**Comment by 0wnrepo** [comment]
^the above comments are horrible. http://chris.beams.io/posts/git-commit/

**Comment by nicosomb** [comment]
@0wnrepo still todo on this PR, we'll squash commits before merging ;-)

**Comment by 0wnrepo** [comment]
@nicosomb ah, ok :+1:

File: src/Wallabag/CommentBundle/Resources/config/services.yml [hunk 15]
```diff
@@ -0,0 +1,4 @@
+services:
+#    wallabag_comment.example:
```
Comment (nicosomb) [inline src/Wallabag/CommentBundle/Resources/config/services.yml]: is this file necessary? Maybe I missed something.

**Comment by nicosomb** [comment]
Doc and baggy implementations added.

File: src/Wallabag/CommentBundle/Resources/config/services.yml (hunk 15 above)
Comment (tcitworld) [inline src/Wallabag/CommentBundle/Resources/config/services.yml]: Nope. Generated automatically.

Comment (j0k3r) [inline src/Wallabag/CommentBundle/Resources/config/services.yml]: Same for the routing.yml

**Comment by j0k3r** [comment]
Before merging that, @nicosomb are you ok if we name _annotation_ in french?

**Comment by nicosomb** [comment]
:+1:

**Comment by nicosomb** [comment]
@tcitworld As I replaced `idComment` with `idAnnotation`, I think that we need to fix that: https://github.com/wallabag/annotator/commit/bf1befd31c50a13842682b3bbaa96b74e2ccf9f6 Right?

File: src/Wallabag/AnnotationBundle/DependencyInjection/WallabagCommentExtension.php [hunk 16]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace Wallabag\AnnotationBundle\DependencyInjection;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\HttpKernel\DependencyInjection\Extension;
+
+class WallabagCommentExtension extends Extension
```
Comment (j0k3r) [inline src/Wallabag/AnnotationBundle/DependencyInjection/WallabagCommentExtension.php]: This should be `WallabagAnnotationExtension `

File: docs/fr/user/annotations.rst [hunk 17]
```diff
@@ -0,0 +1,25 @@
+Commentaires
```
Comment (j0k3r) [inline docs/fr/user/annotations.rst]: Documentation needs update too

File: docs/en/user/annotations.rst [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+   :alt: Write your comment
+   :align: center
+
+The text is now highlighted and you can read your comment if you move the mouse cursor over it.
+
+.. image:: ../../img/user/annotations_3.png
+   :alt: Read your comment
+   :align: center
+
+You can create as many comments as you wish.
```
Comment (j0k3r) [inline docs/en/user/annotations.rst]: All this _comment_ should be _annotate_ or _annotation_

File: docs/fr/user/annotations.rst [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+Ensuite, écrivez votre annotation :
+
+.. image:: ../../img/user/annotations_2.png
+   :alt: Write your comment
+   :align: center
+
+Le texte est maintenant surligné et vous pouvez lire le annotation en le survolant avec votre souris.
+
+.. image:: ../../img/user/annotations_3.png
+   :alt: Read your comment
```
Comment (j0k3r) [inline docs/fr/user/annotations.rst]: And you can update the alt too

**Comment by tcitworld** [comment]
Yup, you need to.

**Comment by j0k3r** [comment]
And voilà, everything is cleanup.
